### ISSUE TITLE
Relay nodes started with tailscale 1.24 do not connect

### ISSUE THREAD

**Issue by suharshs** [tailscale/tailscale#4541]
### What is the issue?

When we try to connect a relay node to our tailscale net using a pre-authenticated auth key, the machine never shows up in the admin console. 

Is there any way for me to pin to an older version to see if the issue is version-specific?

### Steps to reproduce

Run the following on our EC2 node:
```
#!/bin/bash

# Enable IP forwarding
echo 'net.ipv4.ip_forward = 1' | tee -a /etc/sysctl.conf
sysctl -p /etc/sysctl.conf

# Install tailscale
yum install yum-utils
yum-config-manager --add-repo https://pkgs.tailscale.com/stable/amazon-linux/2/tailscale.repo
yum --disablerepo="*" --enablerepo=tailscale-stable install -y tailscale
systemctl enable --now tailscaled
tailscale up --advertise-routes={{.SubnetRoutes}} --authkey={{.AuthKey}}
```

### OS

Linux

### OS version

Amazon Linux 2

### Tailscale version

1.24

### Bug report

BUG-6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4bf581589-20220426231557Z-64009afe997b690b

**Comment by bradfitz** [comment]
Got a `tailscale bugreport` from that machine?

Otherwise the Linux kernel dmesg / boot logs don't really tell me much.

**Comment by suharshs** [comment]
Sorry about that, had some issues getting the bugreport, but got it now:
```
BUG-6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4bf581589-20220426231557Z-64009afe997b690b
```

**Comment by suharshs** [comment]
One more data point: output from `tailscale ip -4`:
```
no current Tailscale IPs; state: NeedsLogin
```

We tried with a fresh reusable Auth Key generated by two Admin users on our team and got this message.

**Comment by bradfitz** [comment]
cc @maisem @danderson

```
2022-04-26 23:15:57.863463438 +0000 UTC: Backend: logs: be:6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4bf581589 fe:
2022-04-26 23:15:57.863763991 +0000 UTC: Switching ipn state NoState -> NeedsLogin (WantRunning=false, nm=false)
2022-04-26 23:15:57.864404006 +0000 UTC: control: authRoutine: state:new; goal=nil paused=false
2022-04-26 23:15:57.864431226 +0000 UTC: control: mapRoutine: state:new
2022-04-26 23:15:57.864459027 +0000 UTC: health("overall"): error: state=NeedsLogin, wantRunning=false
2022-04-26 23:15:57.865123751 +0000 UTC: blockEngineUpdates(true)
2022-04-26 23:15:57.865538904 +0000 UTC: wgengine: Reconfig: configuring userspace wireguard config (with 0/0 peers)
2022-04-26 23:15:57.865882655 +0000 UTC: wgengine: Reconfig: configuring router
2022-04-26 23:15:57.866198612 +0000 UTC: wgengine: Reconfig: configuring DNS
2022-04-26 23:15:57.866493971 +0000 UTC: dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:0}
2022-04-26 23:15:57.866815073 +0000 UTC: dns: Resolvercfg: {Routes:{} Hosts:0 LocalDomains:[]}
2022-04-26 23:15:57.86713424 +0000 UTC: dns: OScfg: {Nameservers:[] SearchDomains:[] MatchDomains:[]}
2022-04-26 23:15:57.867159186 +0000 UTC: wgengine: Reconfig done
2022-04-26 23:15:57.878403321 +0000 UTC: user bugreport: BUG-6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4bf581589-20220426231557Z-64009afe997b690b
2022-04-26 23:15:57.890321038 +0000 UTC: ipnserver: conn3: incoming control connection
2022-04-26 23:15:57.890365448 +0000 UTC: ipnserver: conn3: connection from userid 0; root has access
2022-04-26 23:15:57.895402709 +0000 UTC: Start
2022-04-26 23:15:57.895801423 +0000 UTC: control: client.Shutdown()
2022-04-26 23:15:57.895819408 +0000 UTC: control: client.Shutdown: inSendStatus=0
2022-04-26 23:15:57.895853019 +0000 UTC: control: authRoutine: context done.
2022-04-26 23:15:57.895866177 +0000 UTC: control: authRoutine: state:new; goal=nil paused=false
2022-04-26 23:15:57.895877195 +0000 UTC: control: authRoutine: quit
2022-04-26 23:15:57.895892719 +0000 UTC: control: mapRoutine: context done.
2022-04-26 23:15:57.895901721 +0000 UTC: control: mapRoutine: state:new
2022-04-26 23:15:57.895913946 +0000 UTC: control: mapRoutine: quit
2022-04-26 23:15:57.895927472 +0000 UTC: control: Client.Shutdown done.
2022-04-26 23:15:57.895945621 +0000 UTC: using backend prefs; created empty state for "_daemon": Prefs{ra=true dns=true want=false routes=[] nf=on Persist=nil}
2022-04-26 23:15:57.898561739 +0000 UTC: generating new machine key
2022-04-26 23:15:57.909407065 +0000 UTC: machine key written to store
2022-04-26 23:15:57.909485455 +0000 UTC: control: HostInfo: {"IPNVersion":"1.24.0-t82b460255-g100961435","BackendLogID":"6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4b
f581589","OS":"linux","OSVersion":"Amazon Linux 2; kernel=4.14.219-161.340.amzn2.x86_64","Desktop":false,"Hostname":"ip-10-0-1-188","GoArch":"amd64","RoutableIPs":["10.0.0.0/24",
"10.0.1.0/24"],"Services":[{"Proto":"tcp","Port":22,"Description":"sshd"},{"Proto":"tcp","Port":111,"Description":"rpcbind"}]}
2022-04-26 23:15:57.909515543 +0000 UTC: Backend: logs: be:6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4bf581589 fe:
2022-04-26 23:15:57.909549642 +0000 UTC: Switching ipn state NoState -> NeedsLogin (WantRunning=true, nm=false)
2022-04-26 23:15:57.909585734 +0000 UTC: blockEngineUpdates(true)
2022-04-26 23:15:57.909638116 +0000 UTC: Reconfig(down): no changes made to Engine config
2022-04-26 23:15:57.909666435 +0000 UTC: control: authRoutine: state:new; goal=nil paused=false
2022-04-26 23:15:57.909683659 +0000 UTC: control: mapRoutine: state:new
2022-04-26 23:15:57.910212468 +0000 UTC: StartLoginInteractive: url=false
2022-04-26 23:15:57.91023755 +0000 UTC: control: client.Login(false, 2)
2022-04-26 23:15:57.91025361 +0000 UTC: control: authRoutine: context done.
2022-04-26 23:15:57.910279489 +0000 UTC: control: authRoutine: state:new; wantLoggedIn=true
2022-04-26 23:15:57.910292261 +0000 UTC: control: direct.TryLogin(token=false, flags=2)
2022-04-26 23:15:57.910314899 +0000 UTC: control: LoginInteractive -> regen=true
2022-04-26 23:15:57.910327157 +0000 UTC: control: doLogin(regen=true, hasUrl=false)
2022-04-26 23:15:58.210449556 +0000 UTC: control: control server key  from https://controlplane.tailscale.com
2022-04-26 23:15:58.210855032 +0000 UTC: control: Generating a new nodekey.
2022-04-26 23:15:58.211368775 +0000 UTC: control: RegisterReq: onode= node=[t0CPc] fup=false
2022-04-26 23:15:58.866339083 +0000 UTC: control: HostInfo: {"IPNVersion":"1.24.0-t82b460255-g100961435","BackendLogID":"6268dd4f8dd411ba427640252bc1894f50a1176e555ae208daaeb4e4b
f581589","OS":"linux","OSVersion":"Amazon Linux 2; kernel=4.14.219-161.340.amzn2.x86_64","Desktop":false,"Hostname":"ip-10-0-1-188","GoArch":"amd64","RoutableIPs":["10.0.0.0/24",
"10.0.1.0/24"],"Services":[{"Proto":"peerapi-dns-proxy","Port":1}]}
2022-04-26 23:15:58.866386996 +0000 UTC: control: cancelMapSafely: synced=false
2022-04-26 23:15:58.866401541 +0000 UTC: control: cancelMapSafely: wrote to channel
2022-04-26 23:15:58.866416315 +0000 UTC: control: mapRoutine: new map needed while idle.
2022-04-26 23:15:58.866447227 +0000 UTC: control: mapRoutine: state:authenticating
2022-04-26 23:16:08.21301839 +0000 UTC: control: TryLogin: register request: Post "https://controlplane.tailscale.com/machine/register": all connection attempts failed (HTTP: con
text deadline exceeded, HTTPS: context deadline exceeded)
2022-04-26 23:16:08.213063955 +0000 UTC: control: sendStatus: authRoutine-report: state:authenticating
2022-04-26 23:16:08.213093364 +0000 UTC: Received error: register request: Post "https://controlplane.tailscale.com/machine/register": all connection attempts failed (HTTP: conte
xt deadline exceeded, HTTPS: context deadline exceeded)
2022-04-26 23:16:08.213112538 +0000 UTC: control: authRoutine: backoff: 9 msec
2022-04-26 23:16:08.223077169 +0000 UTC: control: authRoutine: state:authenticating; wantLoggedIn=true
2022-04-26 23:16:08.223102171 +0000 UTC: control: direct.TryLogin(token=false, flags=2)
2022-04-26 23:16:08.22312002 +0000 UTC: control: LoginInteractive -> regen=true
2022-04-26 23:16:08.223131126 +0000 UTC: control: doLogin(regen=true, hasUrl=false)
2022-04-26 23:16:08.223141656 +0000 UTC: control: Generating a new nodekey.
```

The line that jumps out suspicious to me right away is:

```
2022-04-26 23:15:58.210449556 +0000 UTC: control: control server key  from https://controlplane.tailscale.com
```

That's from:

```go
	if serverKey.IsZero() {
		keys, err := loadServerPubKeys(ctx, c.httpc, c.serverURL)
		if err != nil {
			return regen, opt.URL, err
		}
		c.logf("control server key %s from %s", serverKey.ShortString(), c.serverURL)

		c.mu.Lock()
		c.serverKey = keys.LegacyPublicKey
		c.serverNoiseKey = keys.PublicKey
		c.mu.Unlock()
		serverKey = keys.LegacyPublicKey
		serverNoiseKey = keys.PublicKey
```

So it's not returning an error, but its `ShortString()` is the empty string? Oh, that `logf` is just busted. Of course it's zero... that's how it got into that block to begin with. It's not assigned until a few lines down.

Also in logs:

```
2022-04-26 23:16:38.386859426 +0000 UTC: control: sendStatus: authRoutine-report: state:authenticating
2022-04-26 23:16:38.387355247 +0000 UTC: Received error: register request: Post "https://controlplane.tailscale.com/machine/register": all connection attempts failed (HTTP: conx)
2022-04-26 23:16:38.387379692 +0000 UTC: control: authRoutine: backoff: 119 msec
2022-04-26 23:16:38.507270789 +0000 UTC: control: authRoutine: state:authenticating; wantLoggedIn=true
2022-04-26 23:16:38.507312125 +0000 UTC: control: direct.TryLogin(token=false, flags=2)
2022-04-26 23:16:38.507335956 +0000 UTC: control: LoginInteractive -> regen=true
2022-04-26 23:16:38.50734697 +0000 UTC: control: doLogin(regen=true, hasUrl=false)
2022-04-26 23:16:38.507360717 +0000 UTC: control: Generating a new nodekey.
2022-04-26 23:16:38.507482792 +0000 UTC: control: RegisterReq: onode= node=[Jjvxy] fup=false
2022-04-26 23:16:48.509384277 +0000 UTC: trying bootstrapDNS("derp9.tailscale.com", "207.148.3.137") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509537824 +0000 UTC: bootstrapDNS("derp9.tailscale.com", "207.148.3.137") for "controlplane.tailscale.com" error: Get "https://derp9.tailscale.com/bootstrapnt
2022-04-26 23:16:48.509554948 +0000 UTC: trying bootstrapDNS("derp9.tailscale.com", "2001:19f0:6401:1d9c:5400:2ff:feef:bb82") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509638075 +0000 UTC: bootstrapDNS("derp9.tailscale.com", "2001:19f0:6401:1d9c:5400:2ff:feef:bb82") for "controlplane.tailscale.com" error: Get "https://derptt
2022-04-26 23:16:48.509651042 +0000 UTC: trying bootstrapDNS("derp8d.tailscale.com", "178.62.44.132") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509706279 +0000 UTC: bootstrapDNS("derp8d.tailscale.com", "178.62.44.132") for "controlplane.tailscale.com" error: Get "https://derp8d.tailscale.com/bootstr-t
2022-04-26 23:16:48.509718763 +0000 UTC: trying bootstrapDNS("derp11.tailscale.com", "2600:1f1e:ee4:5611:ec5c:1736:d43b:a454") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509768594 +0000 UTC: bootstrapDNS("derp11.tailscale.com", "2600:1f1e:ee4:5611:ec5c:1736:d43b:a454") for "controlplane.tailscale.com" error: Get "https://der1t
2022-04-26 23:16:48.509788809 +0000 UTC: trying bootstrapDNS("derp9c.tailscale.com", "155.138.243.219") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509855605 +0000 UTC: bootstrapDNS("derp9c.tailscale.com", "155.138.243.219") for "controlplane.tailscale.com" error: Get "https://derp9c.tailscale.com/bootsat
2022-04-26 23:16:48.509868136 +0000 UTC: trying bootstrapDNS("derp3.tailscale.com", "2400:6180:0:d1::67d:8001") for "controlplane.tailscale.com" ...
2022-04-26 23:16:48.509938083 +0000 UTC: bootstrapDNS("derp3.tailscale.com", "2400:6180:0:d1::67d:8001") for "controlplane.tailscale.com" error: Get "https://derp3.tailscale.cobt
2022-04-26 23:16:48.51057302 +0000 UTC: control: TryLogin: register request: Post "https://controlplane.tailscale.com/machine/register": all connection attempts failed (HTTP: dl)
2022-04-26 23:16:48.5105928 +0000 UTC: control: sendStatus: authRoutine-report: state:authenticating
2022-04-26 23:16:48.510608505 +0000 UTC: Received error: register request: Post "https://controlplane.tailscale.com/machine/register": all connection attempts failed (HTTP: diat)
```

Does this machine just not have working DNS?

**Comment by bradfitz** [comment]
@suharshs, do you have any outbound firewalls on that machine or its cloud network?

**Comment by suharshs** [comment]
thanks @bradfitz. We set up a VPC in our account with public and private subnets as
per: https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Scenario2.html.
* By default, all network ingress and egress traffic from the VPC is turned off.
* We create 2 CIDR blocks per AZ - one public and one private
* We create a NAT gateway in the public subnet and route internet-bound traffic
from the private subnet via the NAT gateway. This is required for outbound
internet access from nodes in the private subnet.
* Nodes in the public subnet have public IPs and connect to the internet
via an internet gateway.
* For each AZ, we run a Tailscale relay node in EC2 in the public subnet,
which advertises routes to the subnet IPs to rest of the VPN.
* We configure 2 Network ACLs in the VPC - one for private subnets and one for
public subnets. In addition to the private subnet ACLs, the public subnet ACL
allows all UDP egress traffic from the subnet (needed for Tailscale) and
inbound TCP traffic on port 443 (for webserver endpoint).

So based on the backend logs do you think its an issue with either the NAT setup or the NACLs?

**Comment by bradfitz** [comment]
Well, it appears that the `tailscaled` process is unable to hit things it needs to hit on TCP port 80 and TCP port 443. Can you try allowing those too? (It doesn't do unencrypted HTTP on port 80... it just uses it to upgrade to something else. But it's supposed to fall back to 443 if 80 seems blocked) In any case, it also needs TCP 443 out regardless.

**Comment by suharshs** [comment]
Thanks, just checked our public NACL and it allows all output traffic.
<img width="1446" alt="image" src="https://user-images.githubusercontent.com/1450614/165434354-9708fe5e-080e-471f-9de2-feb1058e2b97.png">

Also to double check I ran the Reachability Analyzer from the instance to our VPC internet gateway, and it was successful.

**Comment by bradfitz** [comment]
Can you run:

```
$ curl -I -v http://controlplane.tailscale.com/
$ curl -I -v https://derp10.tailscale.com/
$ ip a
$ ip r
```

**Comment by suharshs** [comment]
Indeed the first curl seems to fail:
```
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[   92.381728] cloud-init[3263]: Dload  Upload   Total   Spent    Left  Speed
[   92.388744] cloud-init[3263]: 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 18.193.255.254...
[   92.399245] cloud-init[3263]: * TCP_NODELAY set
[   92.573042] cloud-init[3263]: *   Trying 2a05:d014:386:201:7200:6340:b22f:7df6...
[   92.582288] cloud-init[3263]: * TCP_NODELAY set
[   92.590091] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:201:7200:6340:b22f:7df6: Network is unreachable
[   92.600135] cloud-init[3263]: *   Trying 2a05:d014:386:203:4535:c15c:9ab:8258...
[   92.607689] cloud-init[3263]: * TCP_NODELAY set
[   92.612873] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:203:4535:c15c:9ab:8258: Network is unreachable
[   92.623019] cloud-init[3263]: *   Trying 2a05:d014:386:203:b7ba:b7d4:9703:6f8a...
[   92.630700] cloud-init[3263]: * TCP_NODELAY set
[   92.636539] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:203:b7ba:b7d4:9703:6f8a: Network is unreachable
[   94.575460] cloud-init[3263]: *   Trying 2a05:d014:386:201:7200:6340:b22f:7df6...
[   94.583645] cloud-init[3263]: * TCP_NODELAY set
[   94.588111] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:201:7200:6340:b22f:7df6: Network is unreachable
[   94.597389] cloud-init[3263]: *   Trying 2a05:d014:386:203:4535:c15c:9ab:8258...
[   94.603962] cloud-init[3263]: * TCP_NODELAY set
[   94.608507] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:203:4535:c15c:9ab:8258: Network is unreachable
[   94.617649] cloud-init[3263]: *   Trying 2a05:d014:386:203:b7ba:b7d4:9703:6f8a...
[   94.624286] cloud-init[3263]: * TCP_NODELAY set
[   94.628692] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:203:b7ba:b7d4:9703:6f8a: Network is unreachable
[   96.577689] cloud-init[3263]: 0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0*   Trying 2a05:d014:386:201:7200:6340:b22f:7df6...
[   96.587091] cloud-init[3263]: * TCP_NODELAY set
[   96.590879] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:201:7200:6340:b22f:7df6: Network is unreachable
[   96.597904] cloud-init[3263]: *   Trying 2a05:d014:386:203:4535:c15c:9ab:8258...
[   96.603606] cloud-init[3263]: * TCP_NODELAY set
[   96.607650] cloud-init[3263]: * Immediate connect fail for 2a05:d014:386:203:4535:c15c:9ab:8258: Network is unreachable
[   96.615587] cloud-init[3263]: *   Trying 2a05:d014:386:203:b7ba:b7d4:9703:6f8a...
```

**Comment by suharshs** [comment]
@bradfitz Thanks for all the help!

I dug around a bit and realized that our NACLs were set up correctly, but the security group only enabled TCP 443 outbound. Although you mentioned that should be enough, I tried also enabling TCP 80 outbound as well and things worked!

This does seem like a potential regression in that fallback behavior since we have had this existing 443 outbound behavior work up until now. (I see you are already tracking that in an issue).

**Comment by bradfitz** [comment]
@suharshs, great! Well, great that there's one fewer mystery. This isn't good but should be fixed by a release later today.

Closing as a dup of since-fixed #4544.
